### ISSUE TITLE
Add event for changes to the safe area

### ISSUE THREAD

**Issue by madsmtm** [rust-windowing/winit#3911]
https://github.com/rust-windowing/winit/pull/3890 added an API to query the safe area, but left an API for listening to changes to that value unspecified (only noting that `RedrawRequested` would be called if it changed).

The API could look something like this:

```rust
enum WindowEvent {
   SafeAreaChanged {
        origin: PhysicalPosition<u32>,
        size: PhysicalSize<u32>,
    }
    // ...
}
```

But that'd mean we have to emit it whenever the surface resizes too, which isn't so nice.

Perhaps something like the following would be better?
```rust
struct Insets {
    top: PhysicalUnit<u32>,
    left: PhysicalUnit<u32>,
    bottom: PhysicalUnit<u32>,
    right: PhysicalUnit<u32>,
}

enum WindowEvent {
   SafeAreaChanged(Insets),
    // ...
}

trait Window {
    fn safe_area_insets(&self) -> Insets;
    // ...
}
```

See also discussion starting from https://github.com/rust-windowing/winit/pull/3890#issuecomment-2343351828.

**Comment by kchibisov** [comment]
>If we expose the insets directly, they would never be wrong, even if the surface itself is resized.

Yeah, that's probably not a bad idea after all, since it's just an _extra padding_ one could say and it'll be much easier actually, since you likely already have a padding handling of some sort in your application just to not make elements render right at the border of the window.

**Comment by madsmtm** [comment]
In https://github.com/rust-windowing/winit/pull/4528, I'm making the safe area on iOS track soft keyboards as well. This has interesting implications for this event, since it means that we need to emit it once per frame while the soft keyboard is opening/closing.

It's not really clear to me how we want this event to interact with `RedrawRequested` though? Do we want safe area changes to also trigger a redraw, or should users do that themselves with `SafeAreaChanged(...) => { window.request_redraw() }`?

WDYT @nicoburns?

**Comment by nicoburns** [comment]
I think probably users do that themselves with `SafeAreaChanged(...) => { window.request_redraw() }`? Not super confident about that. But my reasoning is that if for some reason the user does *not* care about the safe area, then automatically triggering a redraw might cause them to have to render extra frames that they otherwise wouldn't.

**Comment by madsmtm** [comment]
It _would_ unfortunately mean that every application that draws using `safe_area` (which _should_ be all applications) also need `SafeAreaChanged(...) => { window.request_redraw() }`.

**Comment by madsmtm** [comment]
Which seems like annoying boilerplate for a fairly rare edge-case?

**Comment by madsmtm** [comment]
Similar to https://github.com/rust-windowing/winit/pull/4506#issuecomment-4078884486, I'm also thinking a bit about how we should handle this if the user want to get just changes to the keyboard insets, or just changes to the device insets? Would we do two events? Or a single merged event?

**Comment by nicoburns** [comment]
> Similar to [#4506 (comment)](https://github.com/rust-windowing/winit/pull/4506#issuecomment-4078884486), I'm also thinking a bit about how we should handle this if the user want to get just changes to the keyboard insets, or just changes to the device insets? Would we do two events? Or a single merged event?

I would be inclined to do one merged event. Partly because I think we may wish to extend this to more inset types at some point (Android certainly seems to have a lot). This would also allow users to have merged logic that queries a different inset depending on their app state.

But I don't think it matter too much from a user perspective, so I think it would also make sense to let this be driven by implementation efficiency, if it matters for that.

**Comment by nicoburns** [comment]
> It _would_ unfortunately mean that every application that draws using `safe_area` (which _should_ be all applications) also need `SafeAreaChanged(...) => { window.request_redraw() }`.

Yeah... I guess for just the safe area it kinda makes more sense to always redraw. For the keyboard avoiding it might not always? (but tbh probably doesn't matter that much if you do?)
